### PR TITLE
added import sites for discovery

### PR DESCRIPTION
**Pull request by vrindagoel-microsoft**

### CODE REVIEW

File: src/Models/UserInput.cs [hunk 1]
```diff
@@ -38,7 +39,8 @@ public UserInput
         public string TenantId { get; }
         public KeyValuePair<string, string> Subscription { get; }
         public KeyValuePair<string, string> ResourceGroupName { get; }
-        public KeyValuePair<string, string> AzureMigrateProjectName { get; }
+        public KeyValuePair<string, string> AzureMigrateProjectName { get; }        
+        public bool IsImportDiscovery { get; }
         public string DiscoverySiteName { get; }
```
Comment (harshrathore-msft) [inline src/Models/UserInput.cs:44]: We should not add a special flag for import discovery.

If you look at https://github.com/Azure/azure-migrate-export/blob/main/src/Forms/ConfigurationForm.cs
GetApplianceSources function in the above form .

what we should do is add "import" as any other check box and add functions to disable and uncheck other check boxes if import is selected.

File: src/Models/UserInput.cs [hunk 2]
```diff
@@ -20,6 +20,7 @@ public UserInput
             Subscription = subscription;
             ResourceGroupName = resourceGroupName;
             AzureMigrateProjectName = azureMigrateProjectName;
+            IsImportDiscovery = isImportDiscovery;
```
Comment (harshrathore-msft) [inline src/Models/UserInput.cs:23]: We should not need an import flag specially.

File: src/Models/UserInput.cs [hunk 3]
```diff
@@ -10,7 +10,7 @@ public class UserInput
         public UserInput
             (
                 string tenantId, KeyValuePair<string, string> subscription, KeyValuePair<string, string> resourceGroupName, KeyValuePair<string, string> azureMigrateProjectName,
-                string discoverySiteName, string assessmentProjectName,
+                bool isImportDiscovery, string discoverySiteName, string assessmentProjectName,
```
Comment (harshrathore-msft) [inline src/Models/UserInput.cs:13]: We should not need an import flag specially.

File: src/Azure-Migrate-Export.csproj [hunk 4]
```diff
@@ -49,6 +49,7 @@
     <Reference Include="System.Xml.Linq" />
     <Reference Include="System.Data.DataSetExtensions" />
     <Reference Include="Microsoft.CSharp" />
+	<PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
     <Reference Include="System.Data" />
```
Comment (harshrathore-msft) [inline src/Azure-Migrate-Export.csproj:53]: I think this is an auto - change, we don't need two versions for Newtonsoft.json, please upgrade all version to 13.0.3 or keep using 13.0.3

File: src/Forms/AzureMigrateExportMainForm.cs [hunk 5]
```diff
@@ -81,6 +79,7 @@ private void SubmitButton_Click(object sender, EventArgs e)
             KeyValuePair<string, string> subscription = ProjectDetailsFormObj.GetSelectedSubscription();
             KeyValuePair<string, string> resourceGroup = ProjectDetailsFormObj.GetSelectedResourceGroupName();
             KeyValuePair<string, string> azureMigrateProject = ProjectDetailsFormObj.GetSelectedAzureMigrateProject();
+            bool isImportDiscovery = ProjectDetailsFormObj.GetDiscoveryPreference();
```
Comment (harshrathore-msft) [inline src/Forms/AzureMigrateExportMainForm.cs:82]: We should not need this IsImport flag.

File: src/Forms/AzureMigrateExportMainForm.cs [hunk 6]
```diff
@@ -97,7 +96,7 @@ private void SubmitButton_Click(object sender, EventArgs e)
             bool assessSqlServicesSeparately = AssessmentSettingsFormObj.IsAssessSqlServicesSeparatelyChecked();
 
             UserInput userInputObj = new UserInput(
-                                                    tenantId, subscription, resourceGroup, azureMigrateProject, discoverySiteName, assessmentProjectName,
+                                                    tenantId, subscription, resourceGroup, azureMigrateProject, isImportDiscovery, discoverySiteName, assessmentProjectName,
```
Comment (harshrathore-msft) [inline src/Forms/AzureMigrateExportMainForm.cs:99]: We should not need the "IsImport" flag.

File: src/Forms/AzureMigrateExportMainForm.cs [hunk 7]
```diff
@@ -866,5 +865,12 @@ public void MakeTrackProgressTabButtonEnableDecisions()
             // Previous state must be retained, enabled tab buttons do not change.
         }
         #endregion
+
+        #region
+        public bool isImportDiscovery()
+        {
+            return ProjectDetailsFormObj.GetDiscoveryPreference();
+        }
```
Comment (harshrathore-msft) [inline src/Forms/AzureMigrateExportMainForm.cs:873]: We should not need this "IsImport"

File: src/Forms/AzureMigrateExportMainForm.cs [hunk 8]
```diff
@@ -60,7 +60,6 @@ private void NextButton_Click(object sender, EventArgs e)
         {
             if (CurrentButtonTab == ProjectDetailsTabButton)
                 HandleTabChange(ConfigurationFormObj, ConfigurationTabButton);
-
```
Comment (harshrathore-msft) [inline src/Forms/AzureMigrateExportMainForm.cs:63]: Thanks for removing extra spaces 👍

File: src/Forms/ConfigurationForm.cs [hunk 9]
```diff
@@ -16,14 +16,26 @@ public ConfigurationForm(AzureMigrateExportMainForm obj)
...
+            {
+                HyperVCheckBox.Enabled = false;
+                VMwareCheckBox.Enabled = false;
+                PhysicalCheckBox.Enabled = false;
+            }
+            else
+            {
             HyperVCheckBox.Checked = true;
             VMwareCheckBox.Checked = true;
             PhysicalCheckBox.Checked = true;
```
Comment (harshrathore-msft) [inline src/Forms/ConfigurationForm.cs:37]: Indentation does not seem correct.

Comment (harshrathore-msft) [inline src/Forms/ConfigurationForm.cs:37]: This entire logic should not be required.

The SetDefaultConfigurationValues is called only once from the Azure Migrate Export main form.

So, at that time, we can set "HYPERV" "PHYSICAL" "VMWARE" enabled and import as disabled.

Then you need to add "Checkbox_Checked()" functions, for all 4 checkboxed, for appliance check boxes, you disable and uncheck import and for import checkbox you should disable and uncheck import.

File: src/Forms/ConfigurationForm.cs [hunk 10]
```diff
@@ -16,14 +16,26 @@ public ConfigurationForm(AzureMigrateExportMainForm obj)
         {
             InitializeComponent();
             mainFormObj = obj;
+
+            // Set the DropDownStyle to DropDownList to disable text entry
+            ModuleComboBox.DropDownStyle = ComboBoxStyle.DropDownList;
```
Comment (harshrathore-msft) [inline src/Forms/ConfigurationForm.cs:21]: Why was this required?

File: src/Forms/ConfigurationForm.cs [hunk 11]
```diff
@@ -103,7 +115,7 @@ private bool ValidateAzureMigrateSourceAppliance()
             if (VMwareCheckBox.Checked == true || HyperVCheckBox.Checked == true || PhysicalCheckBox.Checked == true)
                 return true;
 
-            return false;
+            return mainFormObj.isImportDiscovery();
```
Comment (harshrathore-msft) [inline src/Forms/ConfigurationForm.cs:118]: This should not be required.

File: src/Forms/ConfigurationForm.cs [hunk 12]
```diff
@@ -223,6 +235,9 @@ public List<string> GetAzureMigrateSourceAppliances()
             if (PhysicalCheckBox.Checked == true)
                 azureMigrateSourceAppliances.Add("physical");
 
+            if (VMwareCheckBox.Checked == false && HyperVCheckBox.Checked == false && PhysicalCheckBox.Checked == false)
+                azureMigrateSourceAppliances.Add("import");
```
Comment (harshrathore-msft) [inline src/Forms/ConfigurationForm.cs:239]: This logic is wrong as per latest mocks,
Just add import if import is checked.

We will handle the disable and enable of appliance checkboxes on the form.

File: src/Discovery/Discover.cs [hunk 13]
```diff
@@ -74,6 +74,60 @@ public bool BeginDiscovery()
             if (UserInputObj.CancellationContext.IsCancellationRequested)
                 UtilityFunctions.InitiateCancellation(UserInputObj);
 
+            int excelCreationPercentProgress = 0;
+            if (UserInputObj.IsImportDiscovery)
+                excelCreationPercentProgress = PerformImportDiscovery(masterSitesObj);
```
Comment (harshrathore-msft) [inline src/Discovery/Discover.cs:79]: This check can be modified to "userInputObj.ApplianceSources" cotaining "import"

File: src/Discovery/Discover.cs [hunk 14]
```diff
@@ -74,6 +74,60 @@ public bool BeginDiscovery()
...
+        private void CreateDiscoveryPropertiesModel(DiscoveryProperties discoveryProperites)
+        {
+            UserInputObj.LoggerObj.LogInformation("Generating model for discovery properties");
+            discoveryProperites.TenantId = UserInputObj.TenantId;
+            discoveryProperites.Subscription = UserInputObj.Subscription.Value;
+            discoveryProperites.ResourceGroup = UserInputObj.ResourceGroupName.Value;
+            discoveryProperites.AzureMigrateProjectName = UserInputObj.AzureMigrateProjectName.Value;
+            discoveryProperites.DiscoverySiteName = UserInputObj.DiscoverySiteName;
+            discoveryProperites.Workflow = UserInputObj.WorkflowObj.IsExpressWorkflow ? "Express" : "Custom - Discovery";
+            discoveryProperites.SourceAppliances = string.Join(",", UserInputObj.AzureMigrateSourceAppliances);
```
Comment (harshrathore-msft) [inline src/Discovery/Discover.cs:126]: If we use "IsImport" flag we will break this functionlatiy.

File: src/Discovery/Discover.cs [hunk 15]
```diff
@@ -176,50 +230,60 @@ public bool BeginDiscovery()
...
-            UserInputObj.LoggerObj.LogInformation("Discovery report found, please ensure the file is closed otherwise deleting it won't be possible and process will terminate");
-            Directory.Delete(directory, true);
-        }
+            try
+            {
+                DiscoveredData.AddRange(importMachinesDiscovery);
+            }
+            catch (Exception exConsolidateVmware)
+            {
+                UserInputObj.LoggerObj.LogWarning($"Failed data consolidation from Import sites: {exConsolidateVmware.Message}");
```
Comment (harshrathore-msft) [inline src/Discovery/Discover.cs:283]: Exception exConsolidateImport

This is not VmWare

File: src/Discovery/RetrieveDiscoveryData.cs [hunk 16]
```diff
@@ -243,6 +259,44 @@ private List<DiscoveryData> ParsePhysicalDiscoveryData(UserInput userInputObj, M
...
+                discoveryDataObj.SqlDiscoveryServerCount = 0;
+                discoveryDataObj.IsSqlServicePresent = false;
+                discoveryDataObj.WebAppCount = 0;
+
+                discoveryDataObj.OperatingSystem = value.Properties.OperatingSystemDetails?.OSName;
+                discoveryDataObj.Cores = value.Properties.NumberOfProcessorCore ?? 0;
+                discoveryDataObj.MemoryInMB = value.Properties.AllocatedMemoryInMB ?? 0.0;
+                discoveryDataObj.TotalDisks = value.Properties.Disks?.Count?? 0;                
+                discoveryDataObj.IpAddress = value.Properties.IPAddresses != null ? string.Join(",", value.Properties.IPAddresses.Select(ip => $"[{ip}]")) + ";" : "";
+                discoveryDataObj.MacAddress = value.Properties.MacAddress;
```
Comment (harshrathore-msft) [inline src/Discovery/RetrieveDiscoveryData.cs:285]: Need to verify this, is it given that there can be only one Mac Address?

**Review by harshrathore-msft — CHANGES_REQUESTED** [review]

File: src/Forms/ConfigurationForm.cs (hunk 10 above)
Comment (vrindagoel-microsoft) [reply src/Forms/ConfigurationForm.cs:21]: because in the custom workflow module box, ideally a user (rather than choosing between the 2 options) shouldn't be allowed to type his own workflow and then receive an error.

File: src/Models/UserInput.cs (hunk 1 above)
Comment (vrindagoel-microsoft) [reply src/Models/UserInput.cs:44]: In the spec document it was written that the appliance section will get disabled or sth, would modify this section according to the recent mocks.

File: src/Forms/ConfigurationForm.cs (hunk 12 above)
Comment (vrindagoel-microsoft) [reply src/Forms/ConfigurationForm.cs:239]: Would modify this section according to the recent mocks.

File: src/Discovery/RetrieveDiscoveryData.cs (hunk 16 above)
Comment (vrindagoel-microsoft) [reply src/Discovery/RetrieveDiscoveryData.cs:285]: verified. It is a string
image

File: src/Azure-Migrate-Export.csproj (hunk 4 above)
Comment (harshrathore-msft) [reply src/Azure-Migrate-Export.csproj]: The package is present Vrinda, please re-verify this.

File: src/Discovery/Discover.cs [hunk 17]
```diff
@@ -74,6 +74,60 @@ public bool BeginDiscovery()
             if (UserInputObj.CancellationContext.IsCancellationRequested)
                 UtilityFunctions.InitiateCancellation(UserInputObj);
 
+            int excelCreationPercentProgress = 0;
+            if (UserInputObj.AzureMigrateSourceAppliances.Contains("import"))
+                excelCreationPercentProgress = PerformImportDiscovery(masterSitesObj);
+            else
+                excelCreationPercentProgress = PerformApplianceDiscovery(masterSitesObj);
```
Comment (harshrathore-msft) [inline src/Discovery/Discover.cs:81]: For this to work properly,

please ensure that we are not sending any other appliance source with import, or add another condition that "import" is the only element in sources.

**Review by harshrathore-msft — APPROVED** [review]
Approved. Good 👍
